### PR TITLE
Add retry options for creating etcd client

### PR DESCRIPTION
**Pull request by robskillington**
When M3 processes using m3cluster for service discovery comes up, a lot of processes will log a fatal error message and fail if they cannot start up.  

This is fine but can cause a large amount of DNS requests to be issued if restarted by a processes supervisor very frequently.  To avoid some of these thundering herd issues if a KV client is down, we add some level of backoff when trying to connect to service discovery.

### CODE REVIEW

File: client/etcd/options.go [hunk 1]
```diff
@@ -30,13 +30,20 @@ import (
...
 
 const (
 	defaultKeepAliveEnabled         = false
 	defaultKeepAlivePeriod          = 30 * time.Minute
 	defaultKeepAlivePeriodMaxJitter = 0
 	defaultKeepAliveTimeout         = 20 * time.Second
+
+	defaultRetryInitialBackoff = 2 * time.Second
+	defaultRetryBackoffFactor  = 2.0
+	defaultRetryMaxBackoff     = 4
```
Comment (xichen2020) [inline client/etcd/options.go]: Should this be `4 * time.Second` rather?

File: client/etcd/options.go [hunk 2]
```diff
@@ -159,17 +166,26 @@ func NewOptions() Options {
...
 		sdOpts: services.NewOptions(),
 		iopts:  instrument.NewOptions(),
+		// NB(r): Set some default retry options so changes to retry
+		// option defaults don't change behavior of this client's retry options
+		retryOpts: retry.NewOptions().
+			SetInitialBackoff(defaultRetryInitialBackoff).
+			SetBackoffFactor(defaultRetryBackoffFactor).
+			SetMaxBackoff(defaultRetryMaxBackoff).
+			SetMaxRetries(defaultRetryMaxRetries).
+			SetJitter(defaultRetryJitter),
```
Comment (xichen2020) [inline client/etcd/options.go]: Do we want to retry forever?

File: client/etcd/options.go (hunk 1 above)
Comment (cw9) [reply client/etcd/options.go]: +1

**Comment by cw9** [comment]
Can we add some details on why we need this in the pr description?

File: client/etcd/options.go (hunk 1 above)
Comment (robskillington) [reply client/etcd/options.go]: Sure thing, I'll just set this to zero.

File: client/etcd/options.go (hunk 2 above)
Comment (robskillington) [reply client/etcd/options.go]: Nah, crash looping the process after 16s is fine.  We just don't want to crash loop the process with no backoff at all.

**Pull request merged**